libr/io/map.c

```diff
@@ -1,372 +1,265 @@
+R_API bool r_io_map_priorize(RIO* io, ut32 id) {
+	SdbListIter* iter;
+	RIOMap* map;
+	if (!io || !io->maps) {
+		return false;
 	}
-	return next;
-}
-
-R_API int r_io_map_del_at(RIO *io, ut64 addr) {
-	RIOMap *map;
-	SdbListIter *iter;
-	ls_foreach_prev (io->maps, iter, map) {
-		if (map->from <= addr && addr < map->to) {
-			ls_delete (io->maps, iter);
+	ls_foreach (io->maps, iter, map) {
+		//search for iter with the correct map
+		if (map->id == id) {
+			ls_split_iter (io->maps, iter);
+			ls_append (io->maps, map);
 			return true;
 		}
 	}
 	return false;
 }
 
-R_API RIOMap *r_io_map_add_next_available(RIO *io, int fd, int flags, ut64 delta, ut64 addr, ut64 size, ut64 load_align) {
+R_API bool r_io_map_priorize_for_fd(RIO* io, int fd) {
+	SdbListIter* iter, * ator;
 	RIOMap *map;
-	SdbListIter *iter;
-	ut64 next_addr = addr,
-		 end_addr = next_addr + size;
-	ls_foreach_prev (io->maps, iter, map) {
-		next_addr = R_MAX (next_addr, map->to+(load_align - (map->to % load_align)));
-		// XXX - This does not handle when file overflow 0xFFFFFFFF000 -> 0x00000FFF
-		// adding the check for the map's fd to see if this removes contention for
-		// memory mapping with multiple files.
-
-		if (map->fd == fd && ((map->from <= next_addr && next_addr < map->to) ||
-			(map->from <= end_addr  && end_addr < map->to)) ) {
-			//return r_io_map_add(io, fd, flags, delta, map->to, size);
-			next_addr = map->to + (load_align - (map->to % load_align));
-			return r_io_map_add_next_available(io, fd, flags, delta, next_addr, size, load_align);
-		} else break;
-	}
-	return r_io_map_new (io, fd, flags, delta, next_addr, size);
-}
-
-R_API RIOMap *r_io_map_add(RIO *io, int fd, int flags, ut64 delta, ut64 addr, ut64 size) {
-	RIOMap *map;
-	SdbListIter *iter;
-	ut64 end_addr = addr + size;
-	if (!flags) {
-		flags = R_IO_READ;
+	SdbList* list;
+	if (!io || !io->maps) {
+		return false;
 	}
-	ls_foreach_prev (io->maps, iter, map) {
-		// XXX - This does not handle when file overflow 0xFFFFFFFF000 -> 0x00000000
-		// keeping (fd, to, from) tuples as separate maps
-		if ( map->fd == fd && ((map->from <= addr && addr < map->to) ||
-			(map->from <= end_addr  && end_addr < map->to)) )
-			//return r_io_map_add(io, fd, flags, delta, map->to, size);
-			return NULL;
+	if (!(list = ls_new ())) {
+		return false;
 	}
-	return r_io_map_new (io, fd, flags, delta, addr, size);
-}
-
-R_API int r_io_map_exists_for_offset (RIO *io, ut64 off) {
-	int res = false;
-	RIOMap *im = NULL;
-	SdbListIter *iter;
-	ls_foreach_prev (io->maps, iter, im) {
-		if (im->from <= off && off < im->to) {
-			res = true;
-			break;
+	//we need a clean list for this, or this becomes a segfault-field
+	r_io_map_cleanup (io);
+	//tempory set to avoid free the map and to speed up ls_delete a bit
+	io->maps->free = NULL;
+	ls_foreach_safe (io->maps, iter, ator, map) {
+		if (map->fd == fd) {
+			ls_prepend (list, map);
+			ls_delete (io->maps, iter);
 		}
 	}
-	return res;
+	while (ls_length (list)) {
```

Comment:
I don't see why they are represented as `SdbList` when serialization is not needed....